File: 백준/Gold/5014. 스타트링크/스타트링크.py
import sys
from collections import deque

input = sys.stdin.readline

F, S, G, U, D = map(int, input().split())

visited = [-1] * (F + 1)
queue = deque()
queue.append(S)
visited[S] = 1

def bfs():
    while queue:
        s = queue.popleft()

        if s == G:
            print(visited[G] - 1)
            return

        for i in (s - D, s + U):
            if 1 <= i <= F and visited[i] == -1:
                visited[i] = visited[s] + 1
                queue.append(i)

    print("use the stairs")
    return

bfs()

File: 백준/Silver/5014. 스타트링크/스타트링크.js
const fs = require("fs");
const stdin = (
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString()
    : fs.readFileSync("./input.txt").toString()
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [F, S, G, U, D] = input()
  .trim()
  .split(" ")
  .map((v) => Number(v));

const building = Array(F + 1).fill(0);
building[S] = 1;
console.log(bfs([[S, 0]]));

function bfs(queue) {
  while (queue.length) {
    const [curr, idx] = queue.shift();
    if (curr === G) return idx;

    for (const next of [curr + U, curr - D]) {
			if (next > 0 && next <= F && !building[next]) {
				building[next] = 1;
        queue.push([next, idx + 1]);
      }
    }
	}
	return "use the stairs";
}
